### ISSUE TITLE
Knowing the number of threads

### ISSUE THREAD

**Issue by mingodad** [h2o/h2o#833]
Hello !
I'm creating a scripting module and it has a debug option that when enabled will reload and compile the script for every request (when doing web development we change a file and expect just to reload the page on the web browser) but for it to work properly h2o need to be configured to only use one thread so I was willing to check it at configuration time to abort/warn the user about it (because the scripting reload/compile will be a mess, one thread compiling and other reloading/invalidating, ...), but actually there is no way to know from the api how many threads are running (if the user do not specify on the config file in main.c it'll be set for the number of cpus but without setting it back to the configurator).

Cheers !

**Comment by kazuho** [comment]
@mingodad 
> h2o need to be configured to only use one thread so I was willing to check it at configuration time to abort/warn the user about it

You can use the `on_context_init` callback to detect if multiple threads are used for handling requests.  The callback exists for handlers to setup per-thread context - you could maintain a mutex-locked counter within `h2o_handler_t` that gets incremented everytime the callback is called, and report configuration error if the value is not zero.

**Comment by mingodad** [comment]
I'm not sure what you propose is a good method !
We can have have multiple hosts on the same server and even if the server is set to 1 thread only probably we'll receive multiple on_context_init for each host/path, or am I missing something here ?

Cheers !

**Comment by kazuho** [comment]
> We can have have multiple hosts on the same server and even if the server is set to 1 thread only probably we'll receive multiple on_context_init for each host/path

Right.  What I am suggesting is that since `on_context_init` would be called for each host/path (_that has different `h2o_handler_t`), you can maintain a counter as a member of `h2o_handler_t` (or a struct that extends it).

> We need a on_thread_start/on_thread_end callback ?

No. Handlers are already possible in capturing thread creation / destroys by setting `on_context_init` / `on_context_dispose` callbacks.

**Comment by mingodad** [comment]
Hello !
Looking at the code for initializing and destroy h2o_context_t I noticed that on h2o_context_init there is a static mutex to serialize h2o_context_init_pathconf_context but on h2o_context_dispose there is no serialization for h2o_context_dispose_pathconf_context, is this intentional or a mistake ?

Cheers !
